src/main/java/com/st/lms/controller/AdminController.java

```diff
@@ -0,0 +1,394 @@
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.st.lms.dto.BkAuthPubDTO;
+import com.st.lms.exception.AlreadyExistsException;
+import com.st.lms.exception.BadRequestException;
+import com.st.lms.exception.NotFoundException;
+import com.st.lms.models.Author;
+import com.st.lms.models.Book;
+import com.st.lms.models.BookLoans;
+import com.st.lms.models.Borrower;
+import com.st.lms.models.LibraryBranch;
+import com.st.lms.models.Publisher;
+import com.st.lms.service.AdminService;
+import com.st.lms.utils.DateCalculations;
+
+@RestController
+public class AdminController {
+	
+	@Autowired
+	AdminService adminService;
+	
+	@GetMapping("/publishers")
+	public List<Publisher> getAllPublishers() {
+		return adminService.getAllPublishers();
+	}
+	
+	@GetMapping("/publisher/{id}")
+	public ResponseEntity<Publisher> getPublisher(@PathVariable int id) throws NotFoundException {
+		Publisher publisher = adminService.getPublisher(id);
+		if(publisher == null) {
+			throw new NotFoundException("Publisher with id=" + id + " not found");
+		}
+		
+		return new ResponseEntity<>(publisher, HttpStatus.OK);
+	}
+	
+	@PostMapping("/publisher")
+	public ResponseEntity<Publisher> addPublisher(@RequestBody Publisher pub) throws AlreadyExistsException {
+		List<Publisher> pubs = adminService.getAllPublishers();
+		for(Publisher p : pubs) {
+			if(p.getPublisherName().equals(pub.getPublisherName())) {
+				throw new AlreadyExistsException("Publisher " + pub.getPublisherName() + " already exists in the database");
+			}
+		}
+		
+		//ok if code gets here
+		adminService.addPublisher(pub);
+		return new ResponseEntity<>(HttpStatus.CREATED);
+	}
+	
+	@PutMapping("/publisher/{id}")
+	public ResponseEntity<Publisher> updatePublisher(@PathVariable int id, @RequestBody Publisher pub) 
+	throws NotFoundException {
+		
+		Publisher publisher = adminService.getPublisher(id);
+		if(publisher == null) {
+			throw new NotFoundException("Update failed. Publisher with id=" + id + " not found");
+		}
+		
+		adminService.updatePublisher(id, pub);
+		return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+	}
+	
+	@DeleteMapping("/publisher/{id}")
+	public ResponseEntity<Publisher> deletePublisher(@PathVariable int id) throws NotFoundException {
+		Publisher publisher = adminService.getPublisher(id);
+		if(publisher == null) {
+			throw new NotFoundException("Delete failed. Publisher with id=" + id + " not found");
+		}
+		
+		adminService.deletePublisher(id);
+		return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+	}
+	
+	/*###################################################################################*/

```

Comment:
I am assuming this is the divide between publisher and book. You may want to make a comment or something more comprehensive.
```suggestion
	/*Book uri section*/
```